File: webmvp/src/lib/songId.ts
/** Derive a URL-safe Firestore document id from a song title. */
export function slugifyTitle(title: string): string {
  return title
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

export function createSongId(): string {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID();
  }

  return `song-${Date.now()}-${Math.random().toString(36).slice(2, 11)}`;
}

/** Prefer explicit id, then ASCII slug, then UUID for non-Latin titles. */
export function resolveSongId(title: string, explicitId?: string | null): string {
  const trimmed = explicitId?.trim();
  if (trimmed) {
    return trimmed;
  }

  const slug = slugifyTitle(title);
  if (slug) {
    return slug;
  }

  return createSongId();
}

File: mobile/lib/domain/song_id.dart
import 'dart:math';

String slugifyTitle(String title) {
  return title
      .toLowerCase()
      .trim()
      .replaceAll(RegExp(r'[^a-z0-9]+'), '-')
      .replaceAll(RegExp(r'^-+|-+$'), '');
}

String createSongId() {
  final random = Random.secure();
  final suffix = List.generate(9, (_) => random.nextInt(36).toRadixString(36))
      .join();
  return 'song-${DateTime.now().millisecondsSinceEpoch}-$suffix';
}

String resolveSongId(String title, {String? explicitId}) {
  final trimmed = explicitId?.trim();
  if (trimmed != null && trimmed.isNotEmpty) {
    return trimmed;
  }
  final slug = slugifyTitle(title);
  if (slug.isNotEmpty) {
    return slug;
  }
  return createSongId();
}
